### ISSUE TITLE
Asphalt v5 "Context stack corruption detected"

### ISSUE THREAD

**Issue by davidbrochart** [asphalt-framework/asphalt#135]
With the following MRE, Asphalt v5 crashes with a `RuntimeError: Context stack corruption detected: context 7fed755bfa80 still has 1 active child context(s)`:
```py
from anyio import TASK_STATUS_IGNORED, create_task_group
from anyio.abc import TaskStatus
from asphalt.core import Component, run_application, start_background_task_factory, start_service_task

async def task1(task_factory, *, task_status: TaskStatus[None] = TASK_STATUS_IGNORED):
    await task_factory.start_task(task2)
    task_status.started()

async def task2(*, task_status: TaskStatus[None] = TASK_STATUS_IGNORED):
    async with create_task_group():
        task_status.started()

class Component0(Component):
    async def start(self):
        task_factory = await start_background_task_factory()
        await start_service_task(lambda: task1(task_factory), name="task")

run_application(Component0)
```

**Comment by agronholm** [comment]
There are 3 contexts here at play:

1. the root context (active in `Component0.start()`)
2. the service task context (active in `task1()`)
3. the context for the task factory task (`task2()`)

The root cause seems to be that, because Asphalt contexts are inherited via contextvars, the `task2` context is inherited from the `task1` context, but since the latter is closed later than the former, Asphalt reports a stack corruption issue because the contexts are closed out of order.

The correct solution would probably to allow contexts to be explicitly inherited from a given parent. In this case, the contexts for tasks spawned from a task factory should be inherited from the same context where the task factory was created in, right?

**Comment by davidbrochart** [comment]
> since the latter is closed later than the former, Asphalt reports a stack corruption issue because the contexts are closed out of order.

It seems to me that it's the opposite: `task1`'s context is closed sooner than `task2`'s?

But yes, creating the task factory inside the service task solves the issue:
```py
from anyio import TASK_STATUS_IGNORED, create_task_group
from anyio.abc import TaskStatus
from asphalt.core import Component, run_application, start_background_task_factory, start_service_task

async def task1(*, task_status: TaskStatus[None] = TASK_STATUS_IGNORED):
    task_factory = await start_background_task_factory()
    await task_factory.start_task(task2)
    task_status.started()

async def task2(*, task_status: TaskStatus[None] = TASK_STATUS_IGNORED):
    async with create_task_group():
        task_status.started()

class Component0(Component):
    async def start(self):
        await start_service_task(task1, name="task")

run_application(Component0)```
```
I'm a bit worried that now one has to think about dependencies between contexts, when we just want to launch tasks in the background.
What if we waited until all child contexts are closed before closing a parent context?

**Comment by agronholm** [comment]
> > since the latter is closed later than the former, Asphalt reports a stack corruption issue because the contexts are closed out of order.
> 
> It seems to me that it's the opposite: `task1`'s context is closed sooner than `task2`'s?

That's exactly what I said. I mentioned task1's context after task2's, and thus task1's context is what "latter" referred to.

> 
> But yes, creating the task factory inside the service task solves the issue:
> 
> ```python
> from anyio import TASK_STATUS_IGNORED, create_task_group
> from anyio.abc import TaskStatus
> from asphalt.core import Component, run_application, start_background_task_factory, start_service_task
> 
> async def task1(*, task_status: TaskStatus[None] = TASK_STATUS_IGNORED):
>     task_factory = await start_background_task_factory()
>     await task_factory.start_task(task2)
>     task_status.started()
> 
> async def task2(*, task_status: TaskStatus[None] = TASK_STATUS_IGNORED):
>     async with create_task_group():
>         task_status.started()
> 
> class Component0(Component):
>     async def start(self):
>         await start_service_task(task1, name="task")
> 
> run_application(Component0)```
> ```
> 
> I'm a bit worried that now one has to think about dependencies between contexts, when we just want to launch tasks in the background. What if we waited until all child contexts are closed before closing a parent context?

This has to do with more than just the exit order. It has to do with what resources are available to each background task. As such, when the service task hosting the task factory starts, it makes a shallow copy of the resources available at the time. All the background tasks should therefore have access to only those resources. Do you agree with me on this?

**Comment by davidbrochart** [comment]
> That's exactly what I said. I mentioned task1's context after task2's, and thus task1's context is what "latter" referred to.

Oh yes, my bad :smile: 

> This has to do with more than just the exit order. It has to do with what resources are available to each background task. As such, when the service task hosting the task factory starts, it makes a shallow copy of the resources available at the time. All the background tasks should therefore have access to only those resources. Do you agree with me on this?

I agree. So the task factory doesn't have the service task `task1` in its context, but the background task `task2` does. Your suggestion:

> The correct solution would probably to allow contexts to be explicitly inherited from a given parent. In this case, the contexts for tasks spawned from a task factory should be inherited from the same context where the task factory was created in, right?

would then work.

**Comment by davidbrochart** [comment]
Fixed in #136.
